src/pmdas/bpf/modules/execsnoop.c

```diff
@@ -0,0 +1,340 @@
+        arg_val[e->args_size-1] = '\0';
+    }
+    arg_val[e->args_size-1] = '\0';
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+    event = data;
+
+    /* TODO: use pcre lib */
+    if (env.name && strstr(event->comm, env.name) == NULL)
+        return;
+
+    /* TODO: use pcre lib */
+    if (env.line && strstr(event->comm, env.line) == NULL)
+        return;
+
+    handle_args(event, 0);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+    pmNotifyErr(LOG_ERR, "Lost %llu events on CPU #%d!", lost_cnt, cpu);
+}
+
+int execsnoop_init(dict *cfg, char *module_name)
+{
+    int err;
+    char *val;
+
+    if ((val = pmIniFileLookup(cfg, module_name, "uid")))
+        env.uid = strtol(val, NULL, 10);
+    if ((val = pmIniFileLookup(cfg, module_name, "max_args")))
+        env.max_args = atoi(val);
+    if ((val = pmIniFileLookup(cfg, module_name, "process_count")))
+        env.process_count = atoi(val);
+    if ((val = pmIniFileLookup(cfg, module_name, "include_failed")))
+        env.fails = atoi(val);
+    if ((val = pmIniFileLookup(cfg, module_name, "command")))
+        env.name = val;
+    if ((val = pmIniFileLookup(cfg, module_name, "line")))
+        env.line = val;
+
+    obj = execsnoop_bpf__open();
+    pmNotifyErr(LOG_INFO, "booting: %s", obj->skeleton->name);
+    if (!obj) {
+        pmNotifyErr(LOG_ERR, "failed to open BPF object");
+        return 1;
+    }
+
+    /* initialize global data (filtering options) */
+    obj->rodata->ignore_failed = !env.fails;
+    obj->rodata->targ_uid = env.uid;
+    obj->rodata->max_args = env.max_args;
+
+    /* initialize global data (filtering options) */
+    obj->rodata->ignore_failed = !env.fails;
+    obj->rodata->targ_uid = env.uid;
+    obj->rodata->max_args = env.max_args;
+
+    err = execsnoop_bpf__load(obj);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to load BPF object: %d", err);
+        return err != 0;
+    }
+
+    err = execsnoop_bpf__attach(obj);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to attach BPF programs");
+        return err != 0;
+    }
+
+    /* setup event callbacks */
+    pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES,
+            handle_event, handle_lost_events, NULL, NULL);
+    if (!pb) {
+        err = -errno;
+        pmNotifyErr(LOG_ERR, "failed to open perf buffer: %d", err);
+        return err != 0;
+    }
+
+    // internal/external instance ids
+    fill_instids(env.process_count, &execsnoop_instances);
+
+    return err != 0;
+}
+
+void execsnoop_shutdown()
+{
+    free(execsnoop_instances);
+    perf_buffer__free(pb);
+    execsnoop_bpf__destroy(obj);
+}
+
+
+void execsnoop_refresh(unsigned int item)
+{
+    /* do nothing */
+}
+
+int execsnoop_fetch_to_atom(unsigned int item, unsigned int inst, pmAtomValue *atom)
+{
+    if (inst == PM_IN_NULL) {
+        return PM_ERR_INST;
+    }
+
+    perf_buffer__poll(pb, PERF_POLL_TIMEOUT_MS);
```

Comment:
This function blocks for `PERF_POLL_TIMEOUT_MS` and then returns the number of perf events of the queue it processed. It can return 0 if no new events were found. In our case, a new event is the execution of a new binary (`execve` syscall).

There is a perf event queue and we have a fixed maximum number of instances in PCP (i.e. we don't want one instance for every single `execve` syscall, that could accumulate to tens of thousands over time). In the bcc PMDA we solved this by having a `cache` queue with a maximum size of `process_count`, which holds the most recent perf events (first in first out), and a background thread which polls the perf event queue and adds new events to the `cache` queue. Then, the PMDA returns items from this queue when asked by the `fetch` callback.

@natoscott, @jasonk000 what approach do you recommend for this PMDA?
I'd suggest to poll for new perf events with a short timeout (I'd set it to less than 100ms, imho 100ms is already too long to block in a callback) in the `refresh` function, and store the events in a C array of the size `process_count`, where each value of this array is one event, and `execsnoop_fetch_to_atom` returns values from `cache_array[inst][item]`. When there are more than `process_count` events processed, it should start overwriting the oldest instance (event) of the `cache_array`.

@smalinux: I think this tool is one of the most difficult libbpf tools to port over to PCP. Quite a challenge for the first module! :100:

Reply:
@andreasgerstmayr Yes, I believe that once I have execsnoop, porting many other tools would be a piece of cake! :)

Reply:
@andreasgerstmayr Done. Thank you.

Reply:
@andreasgerstmayr Done. Thanks.